### ISSUE TITLE
cudaErrorNoKernelImageForDevice does not produce async error

### ISSUE THREAD

**Issue by psalz** [AdaptiveCpp/AdaptiveCpp#1096]
**Bug summary**
When compiling for a CUDA target not supported by the available hardware, kernels are simply not executed, without any notification.

**To Reproduce**
When I compile this program for `sm_86` and run it on a RTX 2070 (`sm_70`), I get no error message whatsoever:

`syclcc --hipsycl-targets="cuda:sm_86" test.cpp -o test && ./test`
```c++
#include <sycl/sycl.hpp>
int main() {
    sycl::queue q{sycl::gpu_selector_v};
    q.parallel_for(sycl::range<1>(10), [](sycl::id<1> idx) {
        printf("hello %lu\n", idx[0]);
    });
    q.wait_and_throw();
}
```

Running through `compute-sanitizer` reveals the issue:

```
compute-sanitizer ./test
========= COMPUTE-SANITIZER
========= Program hit cudaErrorNoKernelImageForDevice (error 209) due to "no kernel image is available for execution on the device" on CUDA API call to cudaLaunchKernel.
```

**Comment by illuhad** [comment]
Yeah, this is a known issue. The core problem is that in the `cuda` and `hip` compilation flows, we never see the `cudaError_t` object returned from `cudaLaunchKernel`, because the `cudaLaunchKernel` call is auto-generated by clang's CUDA/HIP driver.
And for whatever reason, the error then does not show up in the CUDA runtime error state either, so I'm not sure how to retrieve this information.

The generic SSCP compiler `--hipsycl-targets=generic` does not have this problem, so if you need better validation you could give it a try.

**Comment by psalz** [comment]
> Yeah, this is a known issue.

[I figured as much](https://github.com/OpenSYCL/OpenSYCL/issues/47#issuecomment-483839291), I just wanted to document it somewhere.

> The generic SSCP compiler `--hipsycl-targets=generic` does not have this problem, so if you need better validation you could give it a try.

I did actually briefly try it for the first time today to see if it would resolve this issue, but it exploded in other ways (on the `sm_86` machine with `"'+ptx86' is not a recognized feature for this target (ignoring feature)"` followed by `"cuda_executable_object: could not load module (error code = CU:218)"`, on the `sm_70` machine just with the latter message).

**Comment by illuhad** [comment]
> I did actually briefly try it for the first time today to see if it would resolve this issue, but it exploded in other ways (on the sm_86 machine with "'+ptx86' is not a recognized feature for this target (ignoring feature)" followed by "cuda_executable_object: could not load module (error code = CU:218)", on the sm_70 machine just with the latter message).

The "not recognized feature" is known, but this should not have any more impact than a warning.

The "could not load module" indicates some kind of JIT failure. Are you by chance compiling with `-g`? Currently nvidia ptx JIT does not like this (see #1062 ). Other causes can be calls to functions that are not yet supported on SSCP, such as `printf` or SYCL 2020 group algorithms.

**Comment by psalz** [comment]
> Other causes can be calls to functions that are not yet supported on SSCP, such as `printf` or SYCL 2020 group algorithms.

Ah yes, I was compiling the above example (which uses `printf`).
